src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -76,23 +87,129 @@ public void completed(Integer result, Void v) {
                             callback.onGetComplete(entry.toCacheEntry(data));
                         }
 
                         @Override
-                        public void failed(Throwable exc, Void v) {
+                        public void failed(Throwable exc, Void ignore) {
                             VolleyLog.e(exc, "Failed to read file %s", file.getAbsolutePath());
+                            try {
+                                afc.close();
+                            } catch (IOException e) {
+                                VolleyLog.e(e, "Failed to close channel after failed read");
+                            }
                             callback.onGetComplete(null);
                         }
                     });
         } catch (IOException e) {
             VolleyLog.e(e, "Failed to read file %s", file.getAbsolutePath());
+            if (channel != null) {
+                try {
+                    channel.close();
+                } catch (IOException e1) {
+                    VolleyLog.e(e1, "Failed to close channel after IOException");
+                }
+            }
             callback.onGetComplete(null);
         }
     }
 
-    /** Returns a file object for the given cache key. */
-    File getFileForKey(String key) {
-        return new File(mRootDirectorySupplier.get(), DiskBasedCacheUtility.getFilenameForKey(key));
+    /** Puts the cache entry with a specified key into the cache. */
+    @Override
+    public void put(final String key, Cache.Entry entry, final OnPutCompleteCallback callback) {
+        if (DiskBasedCacheUtility.wouldBePruned(
+                mTotalSize, entry.data.length, mMaxCacheSizeInBytes)) {
+            return;
+        }
+
+        final File file = DiskBasedCacheUtility.getFileForKey(key, mRootDirectorySupplier);
+        Path path = Paths.get(file.getPath());
+
+        // channel we can close after IOException
+        AsynchronousFileChannel channel = null;
+        try {
+            final AsynchronousFileChannel afc =
+                    AsynchronousFileChannel.open(
+                            path, StandardOpenOption.WRITE, StandardOpenOption.CREATE);
+            channel = afc;
+            final CacheHeader header = new CacheHeader(key, entry);
+            int headerSize = header.getHeaderSize();
+            int size = entry.data.length + headerSize;
+            ByteBuffer buffer = ByteBuffer.allocate(size);
+            header.writeHeader(buffer);
+            buffer.put(entry.data);
+            buffer.flip();
+            afc.write(
+                    /* source= */ buffer,
+                    /* position= */ 0,
+                    /* attachment= */ null,
+                    new CompletionHandler<Integer, Void>() {
+                        @Override
+                        public void completed(Integer resultLen, Void ignore) {
+                            header.size = resultLen;
+                            mTotalSize =
```

Comment:
It looks a little peculiar to have:

mTotalSize = x;
mTotalSize = y;

rather than just
x;
mTotalSize = y;

since the first set is immediately clobbered by the second.

That seems worth fixing, but it does actually raise a question that I don't think we've considered yet. What thread safety guarantees should this cache provide? The old cache had "synchronized" keywords over all public API methods, and since the methods are blocking, that ensures thread safety by ensuring only one operation can occur at a time on the entire cache. The new implementation doesn't appear to do anything to address thread safety, and synchronized keywords wouldn't be enough since the disk operations are now asynchronous. I also think we need to be careful about when we try to grab the lock - if we try to grab it before triggering an async operation, and the lock will be blocked on some other async operation, then we effectively make the non-blocking API blocking. So we'll need to think through solutions here.

I wouldn't block this PR on it - just something we'll have to address in a follow-up.

Reply:
The first value is used to compute the second.